### PR TITLE
Resource policy

### LINKED ISSUE #6789: API-G resource policies require manual removal
# Bug Report

## Description

Deleting an API-G resource policy from the stack configuration doesn't remove the previously defined policy from the endpoint.  

Similar or dependent issues:

- #4926

### PR DESCRIPTION
**Pull request by brianjd**
## What did you implement

Provides default apig resource policy.  This ensures any previously defined resource policy gets overwritten.  

Closes #6789 

## How can we verify it

First create a resource policy that limits traffic to your source ip address.
```yaml
provider:
  name: aws
  runtime: nodejs12.x

  resourcePolicy:
    - Effect: Allow
      Principal: '*'
      Action: execute-api:Invoke
      Resource:
        - execute-api:/*/*/*
      Condition:
        IpAddress:
          aws:SourceIp:
            - '123.123.123.123'
```

Now remove said policy, and test accessing the endpoint using a different source address from original policy.

```yaml
provider:
  name: aws
  runtime: nodejs12.x

 # resourcePolicy:
 #  - Effect: Allow
 #     Principal: '*'
 #    Action: execute-api:Invoke
 #     Resource:
 #       - execute-api:/*/*/*
 #     Condition:
 #       IpAddress:
 #         aws:SourceIp:
 #           - '123.123.123.123' 
```

## Todos

<details>
<summary>Useful Scripts</summary>

- `npm run test-ci` --> Run all validation checks on proposed changes
- `npm run lint-updated` --> Lint all the updated files
- `npm run lint:fix` --> Automatically fix lint problems (if possible)
- `npm run prettier-check-updated` --> Check if updated files adhere to Prettier config
- `npm run prettify-updated` --> Prettify all the updated files

</details>

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js [hunk 1]
```diff
@@ -4,6 +4,22 @@ const _ = require('lodash');
 const BbPromise = require('bluebird');
 
 module.exports = {
+  resourcePolicy: {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js]: Let's not expose it on `module.exports` but keep as local variable

File: lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js [hunk 2]
```diff
@@ -63,6 +79,15 @@ module.exports = {
           Policy: policy,
         }
       );
+    } else {
+      _.merge(
+        this.serverless.service.provider.compiledCloudFormationTemplate.Resources[
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js]: So if I understand right, the workaround is to ensure policy in all cases, and when user didn't provide any, setup one that has no restrictions.

If it's the case, it'll be good to add a comment to explain why we're doing that.

File: lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js (hunk 1 above)
Comment (brianjd) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js]: @medikoo I exposed it on `module.exports` to easily make it available to the test and also to reference locally.  So make it local and have a duplicate policy in the test?

File: lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js (hunk 2 above)
Comment (brianjd) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/restApi.js]: @medikoo You are correct.  I will add a comment to explain.

**Comment by medikoo** [comment]
Unfortunately we need to revert this. It appears that adding "default" policy is not equal when not adding a policy at all.

It was exposed in our integration tests: https://travis-ci.org/serverless/serverless/jobs/620628266

Rvert PR is here: https://github.com/serverless/serverless/pull/7048

@brianjd if you find the way to ensure that attaching default policy will not differ from not attaching any policy at all, we'll be happy to re-review the PR.

**Comment by brianjd** [comment]
@medikoo @pmuens I actually put the wrong mask, not sure what I was thinking putting a /32. Creating a new PR with corrected mask and also added IPv6.

**Pull request merged**